File: app/src/main/java/com/kaltura/kflow/ui/LoginFragment.java
package com.kaltura.kflow.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.google.android.material.textfield.TextInputEditText;
import com.kaltura.client.services.OttUserService;
import com.kaltura.client.types.LoginResponse;
import com.kaltura.client.utils.request.RequestBuilder;
import com.kaltura.client.utils.response.base.ApiCompletion;
import com.kaltura.kflow.R;
import com.kaltura.kflow.manager.PreferenceManager;
import com.kaltura.kflow.ui.debug.DebugFragment;
import com.kaltura.kflow.ui.main.MainActivity;
import com.kaltura.kflow.manager.PhoenixApiManager;
import com.kaltura.kflow.utils.Utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Created by alex_lytvynenko on 11/18/18.
 */
public class LoginFragment extends DebugFragment implements View.OnClickListener {

    private TextInputEditText mUsername;
    private TextInputEditText mPassword;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_login, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ((MainActivity) requireActivity()).getSupportActionBar().setTitle("Login");

        mUsername = getView().findViewById(R.id.username);
        mPassword = getView().findViewById(R.id.password);
        getView().findViewById(R.id.login).setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        Utils.hideKeyboard(getView());
        makeLoginRequest(mUsername.getText().toString(), mPassword.getText().toString());
    }

    private void makeLoginRequest(String email, String password) {
        if (Utils.hasInternetConnection(requireContext())) {
            RequestBuilder requestBuilder = OttUserService.login(PreferenceManager.getInstance(requireContext()).getPartnerId(), email, password,
                    null, Utils.getUUID(requireContext()))
                    .setCompletion((ApiCompletion<LoginResponse>) result -> {
                        if (result.isSuccess()) {
                            PreferenceManager.getInstance(requireContext()).saveKs(result.results.getLoginSession().getKs());
                            PhoenixApiManager.getClient().setKs(result.results.getLoginSession().getKs());
                        }
                    });
            clearDebugView();
            PhoenixApiManager.execute(requestBuilder);
        } else {
            Toast.makeText(requireContext(), "No Internet connection", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Utils.hideKeyboard(getView());
        PhoenixApiManager.cancelAll();
    }

    @Override
    protected int getDebugViewId() {
        return R.id.debug_view;
    }
}

File: app/src/main/java/com/kaltura/kflow/presentation/login/LoginFragment.kt
package com.kaltura.kflow.presentation.login

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.lifecycleScope
import com.kaltura.client.types.StringValue
import com.kaltura.kflow.R
import com.kaltura.kflow.databinding.FragmentLoginBinding
import com.kaltura.kflow.presentation.base.SharedTransitionFragment
import com.kaltura.kflow.presentation.debug.DebugView
import com.kaltura.kflow.presentation.extension.*
import com.kaltura.kflow.presentation.main.Feature
import com.kaltura.kflow.utils.getUUID
import org.koin.androidx.viewmodel.ext.android.viewModel

/**
 * Created by alex_lytvynenko on 11/18/18.
 */
class LoginFragment : SharedTransitionFragment(R.layout.fragment_login) {

    private val viewModel: LoginViewModel by viewModel()

    override val feature = Feature.LOGIN
    private var _binding: FragmentLoginBinding? = null
    private val binding get() = _binding!!
        override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        _binding = FragmentLoginBinding.inflate(inflater,container,false)
        val view = binding.root
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.login.setOnClickListener {
            hideKeyboard()
            makeLoginRequest(binding.username.string, binding.password.string, binding.extraParamsKey.string,
                binding.extraParamsDescription.string, binding.extraParamsValue.string)
        }
        binding.addExtraParams.setOnClickListener { showExtraParams(true) }
        binding.removeExtraParams.setOnClickListener { showExtraParams(false) }

        binding.username.string = viewModel.getSavedUsername()
        binding.password.string = viewModel.getSavedPassword()
    }

    override fun subscribeUI() {
        observeResource(viewModel.loginRequest,
                error = { binding.login.error(lifecycleScope) },
                success = { binding.login.success(lifecycleScope) }
        )
    }

    private fun makeLoginRequest(email: String, password: String, extraParamsKey: String,
                                 extraParamsDescription: String, extraParamsValue: String) {
        withInternetConnection {
            clearDebugView()
            clearInputLayouts()
            if (email.isEmpty()) {
                binding.usernameInputLayout.showError("Empty username")
                return@withInternetConnection
            }
            if (password.isEmpty()) {
                binding.passwordInputLayout.showError("Empty password")
                return@withInternetConnection
            }

            var extraParams: HashMap<String, StringValue>? = null
            if (extraParamsKey.isNotEmpty() && extraParamsDescription.isNotEmpty() && extraParamsValue.isNotEmpty())
                extraParams = hashMapOf(extraParamsKey to StringValue().apply {
                    description = extraParamsDescription
                    value = extraParamsValue
                })

            binding.login.startAnimation {
                viewModel.makeLoginRequest(email, password, getUUID(), extraParams)
            }
        }
    }

    private fun showExtraParams(isShow: Boolean) {
        binding.addExtraParams.visibleOrGone(isShow.not())
        binding.removeExtraParams.visibleOrGone(isShow)
        binding.extraParamsKeyInputLayout.visibleOrGone(isShow)
        binding.extraParamsDescriptionInputLayout.visibleOrGone(isShow)
        binding.extraParamsValueInputLayout.visibleOrGone(isShow)

        if (isShow.not()) {
            binding.extraParamsKey.text?.clear()
            binding.extraParamsDescription.text?.clear()
            binding.extraParamsValue.text?.clear()
        }
    }

    private fun clearInputLayouts() {
        binding.usernameInputLayout.hideError()
        binding.passwordInputLayout.hideError()
        binding.extraParamsKeyInputLayout.hideError()
        binding.extraParamsDescriptionInputLayout.hideError()
        binding.extraParamsValueInputLayout.hideError()
    }
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}